ui/webui/resources/wasm/candle_embedding_gemma/src/lib.rs

```diff
@@ -3,4 +3,986 @@
+            match self.pool {
+                Pool::Cls | Pool::LastToken | Pool::Splade => {
+                    unreachable!("Only Mean Pooling is supported for Gemma3, neither CLS, nor Last-Token, nor SPLADE");
+                }
+                Pool::Mean => {
+                    if batch_size > 1 {
+                        let results: candle_core::Result<Vec<Tensor>> = batch
+                            .pooled_indices
+                            .iter()
+                            .map(|&i| {
+                                let i = i as usize;
+                                let length = input_lengths[i];
+                                let embeddings = outputs.i((i, ..length))?;
+                                embeddings.sum_keepdim(0)? / (length as f64)
+                            })
+                            .collect();
+
+                        Some(Tensor::cat(&results?, 0)?)
+                    } else {
+                        let length = input_lengths[0];
+                        let embeddings = outputs.i((0, ..length))?;
+                        Some((embeddings.sum_keepdim(0)? / (length as f64))?)
+                    }
+                }
+            }
+        } else {
+            None
+        };
+
+        let raw_embeddings = if has_raw_requests {
+            if batch_size > 1 && has_pooling_requests {
+                let mut final_embeddings = Vec::new();
+                for &i in &batch.raw_indices {
+                    let i = i as usize;
+                    let length = input_lengths[i];
+                    final_embeddings.push(outputs.i((i, ..length))?);
+                }
+                Some(Tensor::cat(&final_embeddings, 0)?)
+            } else {
+                // Single batch or all raw requests
+                if batch_size == 1 {
+                    let length = input_lengths[0];
+                    Some(outputs.i((0, ..length))?)
+                } else {
+                    // Multiple sequences, all raw
+                    let mut all_embeddings = Vec::new();
+                    for (i, &length) in input_lengths.iter().enumerate().take(batch_size) {
+                        all_embeddings.push(outputs.i((i, ..length))?);
+                    }
+                    Some(Tensor::cat(&all_embeddings, 0)?)
+                }
+            }
+        } else {
+            None
+        };
+
+        Ok((pooled_embeddings, raw_embeddings))
+    }
+
+    pub fn embed(&self, input_ids: &Tensor) -> candle_core::Result<Tensor> {
+        // Convert a single input tensor into a Batch
+        let vec_ids: Vec<u32> = input_ids.to_vec1::<u32>()?;
+        let batch = Batch {
+            input_ids: vec_ids.clone(),
+            token_type_ids: vec![0; vec_ids.len()],
+            position_ids: (0..vec_ids.len() as u32).collect(),
+            cumulative_seq_lengths: vec![0, vec_ids.len() as u32],
+            max_length: vec_ids.len() as u32,
```

Comment:
if you `let vec_ids_len = vec_ids.len();` prior to creating the `Batch`, you can skip the `.clone()` call